Tasks/VsTest/settingshelper.ts

```diff
@@ -150,87 +147,83 @@ export async function updateSettingsFileAsRequired(settingsFile: string, isParal
+                result = updateRunSettingsWithDataCollector(result, testImpactFriendlyName, testImpactCollectorNode);
+            } else {
+                tl.debug('Enabling test impact data collection by creating new runsettings.')
+                settingsExt = runSettingsExtension;
+                result = CreateSettings(runSettingsTemplate);
+                result = updateRunSettingsWithDataCollector(result, testImpactFriendlyName, testImpactCollectorNode);
+            }
         }
-    }
 
-    if (isDistributedRun && tiaConfig.tiaEnabled) {
-        let baseLineRunId = utils.Helper.readFileContentsSync(tiaConfig.baseLineBuildIdFile, 'utf-8');
-        if (settingsExt === testSettingsExtension) {
-            tl.debug('Enabling tia in testsettings.');
-            result = setupTestSettingsWithTestImpactOn(result, baseLineRunId);
-        } else if (settingsExt === runSettingsExtension) {
-            tl.debug('Enabling tia in runsettings.');
-            result = setupRunSettingsWithTestImpactOn(result, baseLineRunId);
+        if (isDistributedRun && tiaConfig.tiaEnabled) {
+            let baseLineRunId = utils.Helper.readFileContentsSync(tiaConfig.baseLineBuildIdFile, 'utf-8');
+            if (settingsExt === testSettingsExtension) {
+                tl.debug('Enabling tia in testsettings.');
+                result = setupTestSettingsWithTestImpactOn(result, baseLineRunId);
+            } else if (settingsExt === runSettingsExtension) {
+                tl.debug('Enabling tia in runsettings.');
+                result = setupRunSettingsWithTestImpactOn(result, baseLineRunId);
+            } else {
+                tl.debug('Enabling tia by creating new runsettings.');
+                settingsExt = runSettingsExtension;
+                var runsettingsWithBaseLineRunId = runSettingsForTIAOn.replace("{0}", baseLineRunId);
+                result = CreateSettings(runsettingsWithBaseLineRunId);
+            }
+        }
+
+        if (result) {
+            let fileName = utils.Helper.writeXmlFileSync(result, settingsFile, settingsExt);
+            return fileName;
         } else {
-            tl.debug('Enabling tia by creating new runsettings.');
-            settingsExt = runSettingsExtension;
-            var runsettingsWithBaseLineRunId = runSettingsForTIAOn.replace("{0}", baseLineRunId);
-            result = await CreateSettings(runsettingsWithBaseLineRunId);
+            tl.debug('Not editing settings file. Using specified file as it is.')
+            return settingsFile;
         }
     }
-
-    if (result) {
-        utils.Helper.writeXmlFile(result, settingsFile, settingsExt)
-            .then(function (filename) {
-                defer.resolve(filename);
-            });
-    } else {
-        tl.debug('Not editing settings file. Using specified file as it is.')
-        defer.resolve(settingsFile);
+    catch (err) {
+        tl.warning(tl.loc('ErrorWhileUpdatingSettings'));
+        return settingsFile;
```

Comment:
as decided, failing the task here instead on continuing with generic run settings here